### PR TITLE
Allow Promises to be returned by a packages deactivate method

### PR DESCRIPTION
**Pull request by damieng**

### CODE REVIEW

**Comment by damieng** [comment]
All tests and package tests now passing for this, would love to get some extra eyes on it @atom/core

File: spec/package-manager-spec.coffee [hunk 1]
```diff
@@ -1145,44 +1165,60 @@ describe "PackageManager", ->
...
+      thrownError = null
 
       waitsForPromise ->
         atom.packages.activatePackage("package-that-throws-on-deactivate")
 
+      waitsForPromise ->
+        try
+          atom.packages.deactivatePackage("package-that-throws-on-deactivate")
+        catch error
+          thrownError = error
```
Comment (smashwilson) [inline spec/package-manager-spec.coffee]: I don't think this `catch` clause will ever be triggered. What you probably want is:

```coffee
waitsForPromise ->
  atom.packages.deactivatePackage("package-that-throws-on-deactivate").catch(e -> error = e)
```

**Review by maxbrunsfeld — COMMENTED** [review]
Looks great so far. I would just like to see a test in `package-manager-spec.coffee` for the new behavior. I'm thinking we'll have a fixture package whose `deactivate` hook returns a promise, and we'll assert that the promise returned by `deactivatePackages` does not resolve until that promise resolves.

File: src/package-manager.js [hunk 2]
```diff
@@ -744,21 +744,21 @@ module.exports = class PackageManager {
...
 
   // Deactivate all packages
-  deactivatePackages () {
-    this.config.transact(() => {
-      this.getLoadedPackages().forEach(pack => this.deactivatePackage(pack.name, true))
-    })
+  async deactivatePackages () {
+    await this.config.transactAsync(() =>
+      Promise.all(this.getLoadedPackages().map(pack => this.deactivatePackage(pack.name, true)))
+    )
```
Comment (smashwilson) [inline src/package-manager.js]: :+1:

File: src/package.coffee [hunk 3]
```diff
@@ -506,15 +506,18 @@ class Package
...
       try
-        @mainModule?.deactivate?()
-        @mainModule?.deactivateConfig?()
-        @mainActivated = false
-        @mainInitialized = false
+        result = Promise.resolve(@mainModule?.deactivate?()).then =>
+          @mainModule?.deactivateConfig?()
+          @mainActivated = false
+          @mainInitialized = false
       catch e
```
Comment (smashwilson) [inline src/package.coffee]: This `catch` block won't catch errors thrown from `deactivate()`. You'd need a `.catch` on the Promise to handle async errors instead:

```coffee
result = Promise.resolve(@mainModule?.deactivate?()).then =>
  @mainModule?.deactivateConfig?()
  @mainActivated = false
  @mainInitialized = false
result = result.catch e =>
  # Async errors from the promise
  console.error "Error deactivating package '#{@name}', e.stack
```

**Review by smashwilson — DISMISSED** [review]
I found a few subtle things related to try/catch and Promise interactions. Overall looking good :+1:

**Pull request merged**